processor.go

```diff
@@ -57,7 +57,7 @@ type Processor struct {
 
 	state *Signal
 
-	ctx    context.Context
+	done   chan struct{}
```

Comment:
instead of storing the context (which is bad practice), we'll store a channel that is closed when the processor is done. The purpose is very similar, but this is much cleaner.